pkg/controller/ps/upgrade.go

```diff
@@ -0,0 +1,316 @@
+)
+
+func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *appsv1.StatefulSet, cr *apiv1alpha1.PerconaServerMySQL) error {
+	log := logf.FromContext(ctx)
+
+	if cr.Spec.Pause {
+		return nil
+	}
+
+	if cr.HAProxyEnabled() && cr.Status.HAProxy.State != apiv1alpha1.StateReady {
+		log.Info("Waiting for HAProxy to be ready before smart update")
+		return nil
+	}
+
+	if cr.RouterEnabled() && cr.Status.Router.State != apiv1alpha1.StateReady {
+		log.Info("Waiting for MySQL Router to be ready before smart update")
+		return nil
+	}
+
+	// sleep to get new sfs revision
+	time.Sleep(time.Second)
+
+	currentSet := sts
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Name:      sts.Name,
+		Namespace: sts.Namespace,
+	}, currentSet)
+	if err != nil {
+		return errors.Wrap(err, "failed to get current sfs")
+	}
+
+	list := corev1.PodList{}
+	if err := r.Client.List(context.TODO(),
+		&list,
+		&client.ListOptions{
+			Namespace:     sts.Namespace,
+			LabelSelector: labels.SelectorFromSet(sts.Labels),
+		},
+	); err != nil {
+		return errors.Wrap(err, "get pod list")
+	}
+	statefulSetChanged := false
+	for _, pod := range list.Items {
+		if pod.ObjectMeta.Labels["controller-revision-hash"] != sts.Status.UpdateRevision {
+			statefulSetChanged = true
+			break
+		}
+	}
+	if !statefulSetChanged {
+		return nil
+	}
+
+	log.Info("statefulSet was changed, run smart update")
+
+	// TODO: check if the backup is running
+
+	running, err := r.isBackupRunning(ctx, cr)
+	if err != nil {
+		log.Error(err, "can't start 'SmartUpdate'")
+		return nil
+	}
+	if running {
+		log.Info("can't start/continue 'SmartUpdate': backup is running")
+		return nil
+	}
+
+	if sts.Status.ReadyReplicas < sts.Status.Replicas {
+		log.Info("can't start/continue 'SmartUpdate': waiting for all replicas are ready")
+		return nil
+	}
+
+	primaryHost, err := r.getPrimaryHost(ctx, cr)
+	if err != nil {
+		return err
+	}
+	idx, err := getPodIndexFromHostname(primaryHost)
+	if err != nil {
+		return err
+	}
+	primPod, err := getMySQLPod(ctx, r.Client, cr, idx)
+	if err != nil {
+		return errors.Wrap(err, "get primary pod")
+	}
+	log.Info("primary pod", "name", primPod.Name)
+
+	sort.Slice(list.Items, func(i, j int) bool {
+		return list.Items[i].Name > list.Items[j].Name
+	})
+
+	// TODO: do this properly
+	waitLimit := int32(2 * 60 * 60) // 2 hours
+	// waitLimit := cr.Spec.MySQL.LivenessProbe.InitialDelaySeconds
+
+	for _, pod := range list.Items {
+		pod := pod
+		if pod.Name == primPod.Name {
+			continue
+		} else {
+			log.Info("apply changes to secondary pod", "pod name", pod.Name)
+			if err := r.applyNWait(ctx, cr, sts, &pod, waitLimit); err != nil {
+				return errors.Wrap(err, "failed to apply changes")
+			}
+		}
+	}
+
+	log.Info("apply changes to primary pod", "pod name", primPod.Name)
+	if err := r.applyNWait(ctx, cr, sts, primPod, waitLimit); err != nil {
+		return errors.Wrap(err, "failed to apply changes")
+	}
```

Comment:
do you see any way to perform smart update without waiting? what I mean is the function can return after applying update and operator can continue updating pods (or waiting the pod become online) in the next reconciliation.

blocking reconciliation up to 2 hours feels wrong to me. I know the logic is from K8SPXC, I don't like it there either. this is especially important if operator manages multiple clusters.

Reply:
https://github.com/percona/percona-server-mysql-operator/pull/405/commits/89b24eaf16bcee00a23b182ddc4d8da4c621c25b